lib/puma/thread_pool.rb

```diff
@@ -213,6 +217,22 @@ def wait_until_not_full
       end
     end
 
+    def wait_for_threads_to_finish(tick_time: 0.001, max_wait_ticks: 3)
+      @mutex.synchronize do
+        max_wait_ticks.times do |tick|
+          return if @shutdown
+
+          # if we do wait for more than busy threads
+          # and what is in backlog, stop
+          break if tick >= busy + @todo.size
+
+          # this will be signaled once a request finishes,
+          # which can happen earlier than tick time
+          @not_full.wait @mutex, tick_time
+        end
+      end
+    end
+
```

Comment:
This is the main meat-and-potatoes. The wait with timeout makes sense to me. Why do we need the "tick" here? What's the logic/reason behind doing this multiple times?

Reply:
@schneems https://gitlab.com/gitlab-com/gl-infra/infrastructure/issues/8334#note_247859173